dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Voltage/SecondaryVoltageControl/Simplified/SecondaryVoltageControl.mo

```diff
@@ -79,40 +86,53 @@ equation
   frozen = FreezingActivated and ((blockedUp and (UpRefPu - UpPu) > 0) or (blockedDown and (UpRefPu - UpPu) < 0));
   switch1.u2 = frozen;
 
+  //level is sampled every tSample seconds to calculate levelDiscrete
+  if time >= tUpdate + tSample then
+    tUpdate = time;
+    levelDiscrete = level;
+  else
+    tUpdate = pre(tUpdate);
```

Comment:
What is the error message when you use a when? tUpdate is in fact a discrete real, have you tired using discrete keyword?

Reply:
Here is the log file containing the error when I try to build the model with a when condition
[SecondaryVoltageControlSimp.log](https://github.com/dynawo/dynawo/files/12804213/SecondaryVoltageControlSimp.log)

Reply:
by the way it works in Dynawo as tUpdate is probably seen as a discrete real anyway

Reply:
You can try 
when (time > pre(tUpdate) + tSample) then
    tUpdate = time;
    levelDiscrete = level;
  end when;

Reply:
The sampling with "when" works only with ida solver. However, with the simplified solver the sampling is done every 20s instead of every 10s.

Reply:
Interesting one!

In the version proposed above by Joy, the problem is on the condition: when time **>** pre(tUpdate). In this case, the condition becomes true one step after time = pre(tUpdate) + tSampling, so with a time step = 10 s, it is logical that the sampling is done every 20 s. When using IDA, we search for the first moment when the condition changes, thus the solver stops at pre(tUpdate) + tSampling + epsilon. Graphically and even reading the logs, one can believe that the change is at pre(tUpdate) + tSampling while it is slighlty delayed by epsilon (and for each recurrence, nevertheless epsilon being so small, it is not visible).
Let's illustrate this with one recurrence: the previous stop was at 10 s. When the simplified solver gets to 20, the condition in the when is not fulfilled. Then it goes to 30s where it is fulfilled, updates the level and so on.
In case of IDA, at 20s, it is not fulfilled too but at 20.000000000000...1 it is and then everything is getting updated.

So a better implementation would be when (time >= pre(tUpdate) + tSample), and that's when it really gets funny :)
With such an implementation tUpdate and the level are only updated at t = 10 s and no longer in the rest of the simulation (with the simplified solver), which seems to be completely unconsistent and a bug. However, if we think about the logical behind a when implementation, it makes completely sense. The when mechanism is based on the idea of a change in the "validity of the condition", ie we monitor the fact that the condition was true/false at the last time step and is now false/true. In our case, with a time step equal to 10 s, the condition remains true from t = 10 s to the end of the simulation... so the root is no longer activated and even if a root change is raised by something else in the system, tUpdate and the level are not updated.
If you take a hMax equal to 5 for example, it works perfectly well and the level and tUpdate are updated as expected.

-> It means that this structure can't work if the sampling time is equal to the time step (at least at first glance). I don't know if we have had cases where the genericAutomaton has this property and how it worked then 
 
I will see if I manage to find a way to handle this that is not too dirty, especially that doesn't break too much the separation between solver and modeler parts

Reply:
we could try
elpased = 0 initially

when (elpased >= tSample) then
elapsed = 0;
levelDiscrete = level;
end when;

Reply:
argh does not work as we still need a tSample to update elapsed...

Reply:
I tried adding another variable:  Types.Time tUpdatePre(start=0);

equation
  tUpdatePre = pre(tUpdate);
  when (time >= tUpdatePre + tSample) then
    levelDiscrete = level;
    tUpdate = time;
  end when;

I got the result I wanted. However, when I put the keyword discrete to tUpdatepre, I get the same result as when we used : (time >= pre(tUpdate) + tSample). I am not sure why it does not work both ways. Also even if tUpdate is continuous, it does not solve the issue. 

I am still trying another method to change the when condition.

Reply:
what about having an additional boolean to the when condition that change through an elsewhen?

Reply:
Not sure this works but it might give ideas :There is potentially the possibility to use the operator sample 

when sample(0, tSample) then
  levelDiscrete = level;
end when;

however I don't know if this is supported currently in dynawo and it would also change a bit the behavior as the update will be at each tSample multiplier instead of being updated accordingly to the current time.